### ISSUE TITLE
Rethink supporting Windows XP

### ISSUE THREAD

**Issue by liigo** [rust-lang/rust#12842]
I've made some investigation recently, and found that, there are only two APIs which are not supported by Windows XP, throughout the repo tree. There are:

- CreateSymbolicLinkW, only used by native::io::file::symlink()
- GetFinalPathNameByHandleW, only used by native::io::file::readlink()

And native::io::file::{symlink,readlink} are never used throughout the repo tree, except std::io::fs::{symlink,readlink} (which are really never used). 

Do we really need including symlink/readlink functionality inside std? How other languages do?

Currently Windows XP is still wildly used in the world, especially in Asia. Its installed volume is much more than Mac OS + Linux + Unix. We give up supporting XP just because of two api absent? Maybe it's wrong.

cc #11950

**update:** readlink is used by rustc::metadata::filesearch.

**update:** This is my branch: https://github.com/liigo/rust/tree/xp 
I just commented out the usage of CreateSymbolicLinkW and GetFinalPathNameByHandleW. This make the new rustc.exe run successfully in Windows XP, ant it works mostly OK (see my latest comment below).

**Comment by vadimcn** [comment]
We aren't even supporting [VIsta](https://github.com/rust-lang/rust/issues/19992) right now.

**Comment by liigo** [comment]
Servo was expected to be run in Windows XP:
[2015-01-01] https://github.com/servo/servo/wiki/Mozlandia-Rust-In-Gecko#other-platforms
[2015-02-23] https://github.com/servo/servo/wiki/Meeting-2015-02-23#more-rust-in-gecko-discussion

**Comment by huonw** [comment]
To be clear, that discussion is not about running the servo rendering engine on windows xp (it doesn't run on any version of windows atm, aiui), but rather some components written in Rust to be incorporated into Gecko. The distinction is somewhat important because the components may be restricted enough to not use any of the major features XP is missing, meaning that reaching level of support needed for that purpose may not be (a) very hard, and (b) may not be enough for many other applications.

**Comment by vadimcn** [comment]
FWIW, here's a list of APIs used by libstd, that are missing on Windows XP:

AcquireSRWLockExclusive
AcquireSRWLockShared
CancelIoEx
ReleaseSRWLockExclusive
ReleaseSRWLockShared
SetFileInformationByHandle
SetThreadStackGuarantee
SleepConditionVariableSRW
TryAcquireSRWLockExclusive
TryAcquireSRWLockShared
WakeAllConditionVariable
WakeConditionVariable

**Comment by ColonelJ** [comment]
Hi, thanks for the list, that's quite useful to have if that's all of them.

Personally I'd like to see a full Rust core toolchain working on XP, since that's the best way to know that things are working reasonably; we can just run the existing test suite.  You could then run Rust compiler in an XP VM if you want.

Problem there of course is LLVM unhelpfully dropping support for XP.  Might be good to keep tabs on what incompatibilities that's trying to bring in, and maybe maintain an XP compatible version of LLVM as well (ought to be a lot easier than trying to backport later).  I reckon the wider open source community would be interested in this when they realise compilers they use no longer function on XP.

Main question I think is whether backwards compatible alternatives to those APIs should just replace whatever's there currently or should be included only when those APIs aren't available (be it just XP or Vista also).  Benefit of dropping them completely is that the same code is then being run and tested on Win7 as on XP/Vista, so things seem more likely to work across the board (and more consistently) and there's no chance of Rust's software architecture for XP needing to be different.  Downside is that use of these APIs might be more efficient and some features/capabilities may be lost.  It might even be required to use certain APIs to get things working in Vista/Win7 which don't exist in XP.

Indeed the best approach might be to avoid new APIs unless they're either non-essential (e.g. a security feature that should be enabled) or actually required for correct functioning on the newer OS since there is no alternative.  N.B. such alternatives may be considerably more complex than 'convenience' APIs involving a lot more manual coding or restructuring; however compatibility is highly valuable and custom code using simpler primitives permits much more flexibility in the implementation (which you definitely can't get if you rely on OS code to do all the work).

Indeed it is possible to write code which outperforms the OS and maybe even avoids any use of expensive system calls at all.  Given this is a compiler for what claims to be a systems programming language, it doesn't make a lot of sense to rely on high level OS functions in the first place, if there are ways to achieve the same within the Rust system core, using low level primitives.  As such, similar arguments may apply to use of certain syscalls/libraries in Linux.

Needless to say, writing a Rust OS becomes easier the more stuff libstd can do without OS support for a particular CPU architecture (and anything that works in both Windows and Linux should work on a Rust OS too).  Synchronisation between threads is definitely something I'd expect to be in that category.

**Comment by alexchandel** [comment]
Please, please consider this. I regularly develop for customers using Windows XP. The situation is pretty common in controls engineering and chemical engineering. The ability to emit executables that run on XP is valuable. Even a separate target for XP that lacked mutexes and condition variables would be mildly useful, although I've read interesting [emulation methods](http://www.cs.wustl.edu/~schmidt/win32-cv-1.html) for condition variables.

**Comment by nomota** [comment]
I think attitude is important. Kicking out XP would be a mistake after all, because XP will survive a lot longer than your expectation, owing to the virtual machines. XP is insecure, yes, but nothing is 100% secure. Warm acceptance is very important, if you want your beloved Rust-lang to become a true descendant of C/C++.

**Comment by liigo** [comment]
Firefox want to integrate some Rust code or library (such as mp4 parser and url parser). But Firefox is still required to be run in Windows XP as before. This is another reason for Rust to support XP.

http://www.reddit.com/r/rust/comments/3ab5bc/land_initial_rust_mp4_parser_and_unit_tests_in/

**Comment by KostaCitrix** [comment]
I'd love to use Rust at work and am currently exploring use-cases where that might be possible. However, I definitely need Windows XP support to work for the projects I'm working on. 

I don't need the compiler/toolchain to work on XP, but the produced executables should definitely run on XP. 

I can totally accept some APIs not being available on XP (not sure how to mark that during compilation though). But having a Hello World program crash at start with "AquireSRWLockExclusive could not be located in kernel32.dll." is bad.

I would be glad if you could you reconsider at least partial support for XP.

**Comment by retep998** [comment]
I will work on adding support for XP soon.

**Comment by KostaCitrix** [comment]
@retep998 Thank you! Can I track this progress somewhere? I would love to be up-to-date on this :)

**Comment by retep998** [comment]
@KostaCitrix I'll keep this issue updated on any work I do towards this.

**Comment by alexcrichton** [comment]
@retep998 awesome! I was also planning on getting XP working soon, so if you have any questions feel free to reach out to me on IRC.

**Comment by pythonesque** [comment]
> Main question I think is whether backwards compatible alternatives to those APIs should just replace whatever's there currently or should be included only when those APIs aren't available (be it just XP or Vista also). Benefit of dropping them completely is that the same code is then being run and tested on Win7 as on XP/Vista, so things seem more likely to work across the board (and more consistently) and there's no chance of Rust's software architecture for XP needing to be different. Downside is that use of these APIs might be more efficient and some features/capabilities may be lost. It might even be required to use certain APIs to get things working in Vista/Win7 which don't exist in XP.

> Indeed the best approach might be to avoid new APIs unless they're either non-essential (e.g. a security feature that should be enabled) or actually required for correct functioning on the newer OS since there is no alternative. N.B. such alternatives may be considerably more complex than 'convenience' APIs involving a lot more manual coding or restructuring; however compatibility is highly valuable and custom code using simpler primitives permits much more flexibility in the implementation (which you definitely can't get if you rely on OS code to do all the work).

I disagree with literally every aspect of this.  Hamstringing modern Rust for the sake of an operating system that has been discontinued would be ludicrous.  Rust is an opportunity to get *rid* of dumb decisions in past languages.

Do what you want on Windows XP, but don't take away my modern APIs.

**Comment by nomota** [comment]
Rust is just a new language - not very plausible to thrive, yet. So Rust needs to support as large community as possible. So giving a simple and easy work-around for XP is necessary. The idea of banning XP completely is a silly and dangerous idea. You don't want your beloved language Rust appears to be obstinate, do you?

**Comment by eddyb** [comment]
@nomota Supporting an obsolete version of an operating system riddled with zero days and receiving no security updates is actually dangerous and maybe even irresponsible.
Besides, PRs are accepted if you want to make something compatible with XP as long as it comes at no perceivable cost for newer versions, there is no "banning", it's just not "necessary" or high-priority.

**Comment by adrientetar** [comment]
> XP is insecure, yes, but nothing is 100% secure.

Killer argument – I think I'll just leave my door open from now on because well, no lock is 100% secure anyway.

**Comment by tshepang** [comment]
AFAIK
- Rust code will not be allowed in Firefox unless Rust supports XP
- We want Rust in Firefox
- Therefore, Rust must support Windows XP

**Comment by Thiez** [comment]
For how long will Firefox continue to support XP? When will a Firefox version be released (meant for users, not developers) that contains rust code?

**Comment by tshepang** [comment]
@Thiez I think FIrefox will continue to support XP indefinitely. As for your question, there is [this tweet](https://twitter.com/rillian/status/611922628031283201):
> I wouldn't expect rust code in release before Firefox 44.

**Comment by Thiez** [comment]
I very much doubt that XP will be supported indefinitely. Over the years support for Windows 95, 98, ME, and 2000 (and of those 2000 was actually a quite decent :) ) has been dropped. XP *will* follow, it's only a matter of time. So given that, is supporting XP really worth it?

**Comment by nomota** [comment]
Because XP seems more likely to survive longer than those who want to extinct it.

**Comment by tshepang** [comment]
I understand the word "indefinitely" to mean that there are no plans to drop support.

Note that XP is still a very large market, and is said to have more users than desktop Linux.

**Comment by petrochenkov** [comment]
@Thiez
>XP will follow, it's only a matter of time.

The question is - how much time.
Looking at graphs at netmarketshare.com I can easily imagine it being quite popular in 2020. In this case the support is probably worth it.

(I was going to look at Vista/XP support later this year, but since @alexcrichton works approximately 100x faster than me, it won't probably be needed.)

**Comment by KostaCitrix** [comment]
First of all: Thank you so much for fixing this! This is awesome! :+1: 

Is there a way to detect at _compile_ time or at _application init_ time whether a panic'ing API would be used? If I have an app that should run on Windows XP, I don't want to have to test _every_ code path to make sure some unsupported API doesn't panic...

Thanks in advance for your consideration! :)

**Comment by steveklabnik** [comment]
> Is there a way to detect at compile time or at application init time whether a panic'ing API would be used? 

Not currently, no.

**Comment by KostaCitrix** [comment]
>>Is there a way to detect at compile time or at application init time whether a panic'ing API would be used?

>Not currently, no.

Is this planned or possible? I think that would be crucial to be able to write code without being "really careful about which APIs to use". If I wanted to be careful, I wouldn't choose Rust ;)

**Comment by steveklabnik** [comment]
Not currently. You may want to open an issue on the RFCs repo and/or write up an RFC with details on how to do such a thing. :)

**Comment by tshepang** [comment]
@KostaCitrix I'm sure I understand this joke:

> If I wanted to be careful, I wouldn't choose Rust.

Do you mean Rust is too young?

**Comment by KostaCitrix** [comment]
>> If I wanted to be careful, I wouldn't choose Rust.

> Do you mean Rust is too young?

I meant "If I wanted to be careful _about which API to call in order to not panic_, I wouldn't choose Rust". I want my rust to "it compiles, let's ship it" ;)

**Comment by anfractuosity** [comment]
I'm waiting for the nightly to try out this patch.

When I run my Rust program on WinXP at the moment currently I get 'the procedure entry point TryAcquireSRWLockExclusive could not be located in the dynamic link library kernel32.dll'

With this patch, I'm wondering, will my program likely just panic out, or may it run.

**Comment by alexcrichton** [comment]
This will take some time to propagate into the nightlies currently. The 32-bit Windows compiler is currently still using MinGW and is unlikely to work on Windows XP (although I have not tested this). The triple I tested was `i686-pc-windows-msvc`, which we currently cannot build a compiler for due to [this LLVM bug](https://llvm.org/bugs/show_bug.cgi?id=23957). Instead you'd need to build a compiler which can *target* `i686-pc-windows-msvc`. We don't currently build nightlies for targets like this either (e.g. we have no android nightlies), but this is also coming soon!

tl;dr; to try out XP support you need to do this:

```
./configure --build=x86_64-pc-windows-msvc --target=i686-pc-windows-msvc
make
./x86_64-pc-windows-msvc/stage2/bin/rustc.exe foo.rs -C link-args="/subsystem:console,5.01"
```

**Comment by anfractuosity** [comment]
Thanks a lot! 

I'll give that a shot, I didn't realise it requires a separate target

**Comment by anfractuosity** [comment]
I'm a tad confused I've ended up with a rustc in the stage2 dir you mention however during compilation I got 

```
cfg: version 1.3.0-dev (216f6645e 2015-06-29)
cfg: build triple x86_64-pc-windows-msvc
cfg: host triples x86_64-pc-windows-msvc
cfg: target triples x86_64-pc-windows-msvc i686-pc-windows-msvc
cfg: non-build target triples i686-pc-windows-msvc
cfg: host for x86_64-pc-windows-msvc is x86_64
cfg: host for i686-pc-windows-msvc is i386
cfg: os for x86_64-pc-windows-msvc is pc-windows-msvc
cfg: os for i686-pc-windows-msvc is pc-windows-msvc
cfg: good valgrind for x86_64-pc-windows-msvc is
cfg: good valgrind for i686-pc-windows-msvc is
cfg: using CC=gcc (CFG_CC)
cfg: disabling valgrind run-pass tests
cfg: no xelatex found, disabling LaTeX docs
cfg: no pandoc found, omitting PDF and EPUB docs
make: *** No rule to make target 'i686-pc-windows-msvc/rt/rust_try_msvc_32.o', needed by 'i686-pc-windows-msvc/rt/rustrt_native.lib'.  Stop.
```

I've tried doing 
```
 ./rustc --target=i686-pc-windows-msvc foo.rs -C link-args="/subsystem:console,5.01"
foo.rs:1:1: 1:1 error: can't find crate for `std`
foo.rs:1 pub fn main(){
         ^
error: aborting due to previous error
```

**Comment by alexcrichton** [comment]
@anfractuosity you probably won't get very far if the build doesn't complete, so I'd start out diagnosing why `rust_try_msvc_32.o` didn't build. Perhaps a reconfigure was needed?

**Comment by alexcrichton** [comment]
Oh wow it looks like I just totally forgot to commit the file! I'll add it in a PR I'm prepping now, but you can add `src/rt/rust_try_msvc_32.ll` with these contents:

```ll
define i8* @rust_try(void (i8*)* %f, i8* %env) {
    call void %f(i8* %env)
    ret i8* null
}
```

**Comment by anfractuosity** [comment]
Cheers!  That works for me.

In msys2 I did, just so I don't forget ;)

```
 export PATH="C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin":$PATH

./rustc --target=i686-pc-windows-msvc  --verbose foo.rs -C link-args="/subsystem:console,5.01" -L  "C:\Program Files (x86)\Windows Kits\8.1\Lib\winv6.3\um\x86" -L "C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\lib"
```

**Comment by anfractuosity** [comment]
Just wondering if anyone can point me in the right direction, I can compile a simple rust program fine, but I'm getting the following error with a more complex one:

```
......
t\debug\deps\c_vec-460b36a551b013a2.lib and object c:\Users\chris\My Documents\G
ithub\hymntoninkasi\target\debug\deps\c_vec-460b36a551b013a2.exp
LINK : error LNK2001: unresolved external symbol _DllMainCRTStartup
c:\Users\chris\My Documents\Github\hymntoninkasi\target\debug\deps\c_vec-460b36a
551b013a2.dll : fatal error LNK1120: 1 unresolved externals

error: aborting due to previous error
   Compiling gcc v0.3.8
Build failed, waiting for other jobs to finish...
Could not compile `c_vec`.
```

**Comment by MrAlert** [comment]
The MinGW target seems to work for Windows XP. Or at least `stdtest-i686-pc-windows-gnu.exe` gets this far:
````
test ascii::tests::test_ascii ... thread '<main>' panicked at 'condition variables not available', src/libstd\sys/windows\c.rs:415
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', src/libcore\result.rs:731
````
Looks like the test infrastructure already relies on condition variables for something.

**Comment by anfractuosity** [comment]
@MrAlert Was that using the latest nightly build then?
